polygeist/lib/Dialect/Polygeist/Utils/TransformUtils.cpp

```diff
@@ -399,3 +402,62 @@ void LoopTools::versionLoop(LoopLikeOpInterface loop,
                             const LoopVersionCondition &versionCond) const {
   LoopVersionBuilder(loop).versionLoop(versionCond);
 }
+
+//===----------------------------------------------------------------------===//
+// SYCL related functions
+//===----------------------------------------------------------------------===//
+
+static constexpr StringRef linkageAttrName = "llvm.linkage";
+bool mlir::isLinkonceODR(FunctionOpInterface func) {
+  if (!func->hasAttr(linkageAttrName))
+    return false;
+  auto attr = dyn_cast<LLVM::LinkageAttr>(func->getAttr(linkageAttrName));
+  assert(attr && "Expecting LLVM::LinkageAttr");
+  return attr.getLinkage() == LLVM::Linkage::LinkonceODR;
+}
+
+bool mlir::isOnlyCalledFromGPUKernel(FunctionOpInterface func) {
```

Comment:
Here is what is confusing to me about this function: every function in a gpu module is called directly or indirectly from a `gpu.func` with kernel attribute (unless is unreachable). So this function usefulness is very limited. Is almost like this function will return true for every function in the gpu module ?

Reply:
The original code this function aims to replace would return true is the function is called directly from a gpu.func  (and should have checked the kernel attribute is present) or is called from a function that itself is directly called from the gpu.func.

Reply:
changed.